### PR TITLE
no longer rendering un-updated render passes

### PR DESCRIPTION
**Pull request by cixzhang**
This probably isn't quite what you meant since it doesn't get rid of called RAF each render. This will avoid the renderPass block unless the following are updated:
- resize (due to JS or CSS modifications, window events, etc)
- camera changes
- texture updates
- data updates

### CODE REVIEW

File: src/renderer.js [hunk 1]
```diff
@@ -115,63 +158,48 @@ Renderer.prototype.resize = function (resolution) {
...
+    if (a.length != b.length) return false;
+
+    var check = true;
+    for (var i = 0; i < a.length; i++) {
+      check = check && equivalent(a[i], b[i], delta);
+    }
+    return check;
+  }
+  return Math.abs(a - b) <= delta;
+}
```
Comment (ptpham) [inline src/renderer.js:205]: I will allow your dark transgression. This time.

Comment (cixzhang) [inline src/renderer.js:205]: Didn't I ask you about this months earlier? :P

File: src/renderer.js [hunk 2]
```diff
@@ -88,18 +97,52 @@ Renderer.prototype.stop = function () {
...
+    uniforms.u_world = m4.copy([], camera.computedMatrix);
+    uniforms.u_worldView = m4.multiply([], uniforms.u_view, uniforms.u_world);
+    uniforms.u_worldViewProjection = m4.multiply([], projection, uniforms.u_worldView);
+
+    this.updated = true;
+  }
+  return this;
+};
+
 Renderer.prototype.resize = function (resolution) {
```
Comment (ptpham) [inline src/renderer.js:146]: What I've been doing for other stuff is resizing the image backing the canvas before every render. I.e. canvas.height = canvas.offsetHeight and canvas.width = canvas.offsetWidth. I'm not sure if this is performant, but it seems pretty reasonable. XD

**Comment by ptpham** [comment]
This is functionally better. It may be possible to simplify, but it's definitely better. =D

File: src/renderer.js (hunk 1 above)
Comment (ptpham) [inline src/renderer.js:205]: Yeah, but like dark transgressions are always dark transgressions. XD

**Pull request merged**